mlflow/java/spark/src/main/scala/org/mlflow/spark/autologging/MlflowAutologEventPublisher.scala

```diff
@@ -0,0 +1,130 @@
+
+import scala.collection.mutable
+import scala.util.control.NonFatal
+
+/**
+  * Object exposing the actual implementation of MlflowAutologEventPublisher.
+  * We opt for this pattern (an object extending a trait) so that we can mock methods of the
+  * trait in testing
+  */
+object MlflowAutologEventPublisher extends MlflowAutologEventPublisherImpl {
+
+}
+
+/**
+ * Trait implementing a publisher interface for publishing events on Spark datasource reads to
+ * a set of listeners. See the design doc:
+ * https://docs.google.com/document/d/11nhwZtj-rps0stxuIioFBM9lkvIh_ua45cAFy_PqdHU/edit for more
+ * details.
+ */
+private[autologging] trait MlflowAutologEventPublisherImpl {
+  private val logger = LoggerFactory.getLogger(getClass)
+
+  private[autologging] var sparkQueryListener: SparkDataSourceListener = _
+  private val ex = new ScheduledThreadPoolExecutor(1)
+  private[autologging] var subscribers: mutable.LinkedHashMap[String, MlflowAutologEventSubscriber] =
+    mutable.LinkedHashMap[String, MlflowAutologEventSubscriber]()
+  private var scheduledTask: ScheduledFuture[_] = _
+
+  def spark: SparkSession = {
+    SparkSession.builder.getOrCreate()
+  }
+
+  // Exposed for testing
+  private[autologging] def getSparkDataSourceListener: SparkDataSourceListener = {
+    new SparkDataSourceListener()
+  }
+
+  // Initialize Spark listener that pulls Delta query plan information & bubbles it up to registered
+  // Python subscribers, along with a GC loop for removing unrespoins
+  def init(gcDeadSubscribersIntervalSec: Int = 1): Unit = synchronized {
+    if (sparkQueryListener == null) {
+      val listener = getSparkDataSourceListener
+      // NB: We take care to set the variable only after adding the Spark listener succeeds,
+      // in case listener registration throws. This is defensive - adding a listener should
+      // always succeed.
+      spark.sparkContext.addSparkListener(listener)
+      sparkQueryListener = listener
+      // Schedule regular cleanup of detached subscribers, e.g. those associated with detached
+      // notebooks
+      val task = new Runnable {
+        def run(): Unit = {
+          unregisterBrokenSubscribers()
+        }
+      }
+      scheduledTask = ex.scheduleAtFixedRate(task, 1, gcDeadSubscribersIntervalSec, TimeUnit.SECONDS)
+    }
+  }
+
+  def stop(): Unit = synchronized {
+    if (sparkQueryListener != null) {
+      spark.sparkContext.removeSparkListener(sparkQueryListener)
+      sparkQueryListener = null
+      while(!scheduledTask.cancel(false)) {
+        Thread.sleep(1000)
+      }
+      subscribers = mutable.LinkedHashMap()
+    }
+  }
+
+  def register(subscriber: MlflowAutologEventSubscriber): String = synchronized {
+    if (sparkQueryListener == null) {
+      throw new RuntimeException("Please call init() before attempting to register a subscriber")
+    }
+    this.synchronized {
+      val uuid = java.util.UUID.randomUUID().toString
+      subscribers.put(uuid, subscriber)
```

Comment:
why do we need uuid? Seems like it is not really used for anything other than addressing broken subcribers?

Can we a) use replId  or b) use sequential id instead?

Reply:
Good idea, makes sense to use replId. Will do that & make sure it's documented that replId is expected to be unique